Firestore/core/src/firebase/firestore/model/field_path.cc

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/model/field_path.h"
+
+#include <algorithm>
+#include <cctype>
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/util/firebase_assert.h"
+#include "absl/strings/str_join.h"
+#include "absl/strings/str_replace.h"
+#include "absl/strings/str_split.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+namespace {
+
+// TODO(varconst): move to C++ equivalent of FSTDocumentKey.{h,cc}
+const char* const kDocumentKeyPath = "__name__";
+
+/**
+ * True if the string could be used as a segment in a field path without
+ * escaping.
+ */
+bool IsValidIdentifier(const std::string& segment) {
+  if (segment.empty()) {
+    return false;
+  }
+  if (segment.front() != '_' && !std::isalpha(segment.front())) {
```

Comment:
`std::isalpha` is locale sensitive and therefore incorrect.

Reply:
Thanks for catching that.

Note the added comment. Strictly speaking, in C++ it's not guaranteed that 'a'-'z' and 'A'-'Z' are contiguous ranges. However, the only system I'm aware of where this expectation is broken is pretty exotic (EBCDIC used on old IBM mainframes), so I think it makes sense to ignore that.